src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpFormatter.cs

```diff
@@ -41,50 +41,88 @@ internal class CSharpFormatter
             _server = languageServer;
-            _projectSnapshotManagerAccessor = projectSnapshotManagerAccessor;
             _filePathNormalizer = filePathNormalizer;
+
+            var type = typeof(CSharpFormattingOptions).Assembly.GetType("Microsoft.CodeAnalysis.CSharp.Indentation.CSharpIndentationService", throwOnError: true);
+            _indentationService = Activator.CreateInstance(type);
+            var indentationService = type.GetInterface("IIndentationService");
+            _getIndentationMethod = indentationService.GetMethod("GetIndentation");
         }
 
         public async Task<TextEdit[]> FormatAsync(
-            RazorCodeDocument codeDocument,
-            Range range,
-            DocumentUri uri,
-            FormattingOptions options,
+            FormattingContext context,
+            Range rangeToFormat,
             CancellationToken cancellationToken,
             bool formatOnClient = false)
         {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (rangeToFormat is null)
+            {
+                throw new ArgumentNullException(nameof(rangeToFormat));
+            }
+
             Range projectedRange = null;
-            if (range != null && !_documentMappingService.TryMapToProjectedDocumentRange(codeDocument, range, out projectedRange))
+            if (rangeToFormat != null && !_documentMappingService.TryMapToProjectedDocumentRange(context.CodeDocument, rangeToFormat, out projectedRange))
             {
                 return Array.Empty<TextEdit>();
             }
 
             TextEdit[] edits;
             if (formatOnClient)
             {
-                edits = await FormatOnClientAsync(codeDocument, projectedRange, uri, options, cancellationToken);
+                edits = await FormatOnClientAsync(context, projectedRange, cancellationToken);
             }
             else
             {
-                edits = await FormatOnServerAsync(codeDocument, projectedRange, uri, options, cancellationToken);
+                edits = await FormatOnServerAsync(context, projectedRange, cancellationToken);
             }
 
-            var mappedEdits = MapEditsToHostDocument(codeDocument, edits);
+            var mappedEdits = MapEditsToHostDocument(context.CodeDocument, edits);
             return mappedEdits;
         }
 
+        public int GetCSharpIndentation(FormattingContext context, int projectedDocumentIndex, CancellationToken cancellationToken)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            // Add a marker at the position where we need the indentation.
+            var changedText = context.CSharpSourceText;
+            var marker = $"{context.NewLineString}#line default{context.NewLineString}#line hidden{context.NewLineString}";
```

Comment:
I tried a bunch of different things to use as markers. But this is the only one I found to work in all cases.

Reply:
Could you go a little bit into why the other ones didn't work? Do they just appear in other scenarios? How did you determine that this whole string wouldn't appear otherwise?

Reply:
> Could you go a little bit into why the other ones didn't work?

Not entirely sure. Roslyn's `IIndentationService` doesn't like it when I provide markers like `/**/` or empty lines etc. The API returns wacky results that don't make any sense. I plan to follow up with them to understand this further.